### PR TITLE
multipart.tracker benchmarks

### PR DESCRIPTION
**Pull request by itai-david**
Benchmarking PostgresDB vs KV-Mem vs KV-Postgres with the following:
* Create operation
* Sequential Get opertations
* Random Get operations
* Sequential Delete operatrions
* Mix of 10% Create, 80% Get and 10% Delete operations
All of the above can run individually to benchmark a specific op, or all together, as shown below

Sample output:
```
$ go test -bench=BenchmarkMultipartsTracker -benchtime=10000x
goos: linux
goarch: amd64
pkg: github.com/treeverse/lakefs/pkg/gateway/multiparts
cpu: 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz
BenchmarkMultipartsTrackerCreate/benchmarkCreateOp_DB-8         	   10000	    422947 ns/op
BenchmarkMultipartsTrackerCreate/benchmarkCreateOp_KVmem-8      	   10000	      1744 ns/op
BenchmarkMultipartsTrackerCreate/benchmarkCreateOp_KVpg-8       	   10000	    343066 ns/op
BenchmarkMultipartsTrackerGetSeq/benchmarkGetOpSeq_DB-8         	   10000	   1002876 ns/op
BenchmarkMultipartsTrackerGetSeq/benchmarkGetOpSeq_KVmem-8      	   10000	     16226 ns/op
BenchmarkMultipartsTrackerGetSeq/benchmarkGetOpSeq_KVpg-8       	   10000	    126026 ns/op
BenchmarkMultipartsTrackerGetRand/benchmarkGetOpRand_DB-8       	   10000	    208291 ns/op
BenchmarkMultipartsTrackerGetRand/benchmarkGetOpRand_KVmem-8    	   10000	      2384 ns/op
BenchmarkMultipartsTrackerGetRand/benchmarkGetOpRand_KVpg-8     	   10000	     34168 ns/op
BenchmarkMultipartsTrackerDelete/benchmarkDeleteOpSeq_DB-8      	   10000	    408299 ns/op
BenchmarkMultipartsTrackerDelete/benchmarkDeleteOpSeq_KVmem-8   	   10000	      1515 ns/op
BenchmarkMultipartsTrackerDelete/benchmarkDeleteOpSeq_KVpg-8    	   10000	    409746 ns/op
BenchmarkMultipartsTrackerMix/benchmarkMixedOps_DB-8            	   10000	    161558 ns/op
BenchmarkMultipartsTrackerMix/benchmarkMixedOps_KVmem-8         	   10000	       733.1 ns/op
BenchmarkMultipartsTrackerMix/benchmarkMixedOps_KVpg-8          	   10000	     86964 ns/op
PASS
ok  	github.com/treeverse/lakefs/pkg/gateway/multiparts	70.690s

```

### CODE REVIEW

**Comment by arielshaqed** [comment]
Hey @itai-david ,

This is some _very_ nice work!  Note that, because all operations are uncontended, it measures single-threaded behaviour of the multipart tracker when nothing else is running.

# A note on `b.N`

By using it to control the size of the dataset, you may be using `b.N` incorrectly.  Perhaps @nopcoder has a better reference than this [blog by Russ Cox](https://dave.cheney.net/2013/06/30/how-to-write-benchmarks-in-go), but benchmarks should run _the exact same thing_ `b.N` times.  If `b.N` controls anything else, then  as `testing` increases `b.N` it will _change the iteration time_ -- and the results will be inaccurate (the blog has an extreme example, where depending on `b.N` causes the benchmark to diverge and never exit).

**tl;dr:** Run the exact same sequence `b.N` times.

# Benchmarking metadata operations

Multipart is probably dominated by time spent on data, this benchmark measures time spent on data.  So almost any result will be good enough here.

OTOH, multipart uploads are generally conducted in parallel, so if you care about time to perform metadata operations you should measure those.

# Suggestion: benchmark (metadata) operations for a full multipart upload

E.g., pretend you're uploading 1000 parts from k workers.  Benchmark the time to perform all needed operations, where each worker runs on a single goroutine.  Then vary k to find interesting behaviour.

**Review by itaiad200 — APPROVED** [review]
Run this on my station too - similar results. Thanks for the really clear benchmarks!

**Comment by itai-david** [comment]
> Hey @itai-david ,
> 
> This is some _very_ nice work! Note that, because all operations are uncontended, it measures single-threaded behaviour of the multipart tracker when nothing else is running.
> 
> # A note on `b.N`
> By using it to control the size of the dataset, you may be using `b.N` incorrectly. Perhaps @nopcoder has a better reference than this [blog by Russ Cox](https://dave.cheney.net/2013/06/30/how-to-write-benchmarks-in-go), but benchmarks should run _the exact same thing_ `b.N` times. If `b.N` controls anything else, then as `testing` increases `b.N` it will _change the iteration time_ -- and the results will be inaccurate (the blog has an extreme example, where depending on `b.N` causes the benchmark to diverge and never exit).
> 
> **tl;dr:** Run the exact same sequence `b.N` times.
> 
> # Benchmarking metadata operations
> Multipart is probably dominated by time spent on data, this benchmark measures time spent on data. So almost any result will be good enough here.
> 
> OTOH, multipart uploads are generally conducted in parallel, so if you care about time to perform metadata operations you should measure those.
> 
> # Suggestion: benchmark (metadata) operations for a full multipart upload
> E.g., pretend you're uploading 1000 parts from k workers. Benchmark the time to perform all needed operations, where each worker runs on a single goroutine. Then vary k to find interesting behaviour.

Thank you for these insights @arielshaqed 
Per your suggestion - added a full flow (Create, Get, Get ..... Delete) with concurrency.
In addition, created #3387 to follow up on the dataset size wrong dependency in b.N

File: pkg/gateway/multiparts/tracker_test.go [hunk 1]
```diff
@@ -235,3 +237,158 @@ func testTrackerCreate(t *testing.T, tracker multiparts.Tracker) {
...
+	keys := createEntriesForTest(b, ctx, tracker, b.N)
+
+	b.ResetTimer()
+	for n := 0; n < b.N; n++ {
+		_, err := tracker.Get(ctx, keys[n])
+		testutil.Must(b, err)
+	}
+}
+
+func benchmarkGetOpRand(b *testing.B, ctx context.Context, tracker multiparts.Tracker) {
```
[the remark this answers is not in the document]
Comment (itai-david) [reply pkg/gateway/multiparts/tracker_test.go]: This is not trivial as initially seems (at least for some of the benchmarks). I opened #3387 to follow up on that

File: pkg/gateway/multiparts/tracker_test.go [hunk 2]
```diff
@@ -235,3 +237,158 @@ func testTrackerCreate(t *testing.T, tracker multiparts.Tracker) {
...
+
+// TODO - identical to a function in pkg/graveler/sstable/writer_test.go
+// refactor to a common "test_helpers" or alike
+func randomStrings(n int) []string {
+	var keys []string
+	for i := 0; i < n; i++ {
+		keys = append(keys, randstr.String(20, "abcdefghijklmnopqrstuvwyz0123456789"))
+	}
+	return keys
+}
```
[the remark this answers is not in the document]
Comment (itai-david) [reply pkg/gateway/multiparts/tracker_test.go]: Done. Renamed as suggested. Comment removed as it was not a relevant documentation, but a `TODO` that is no longer relevant

**Review by arielshaqed — COMMENTED** [review]
Thanks!

File: pkg/gateway/multiparts/tracker_test.go [hunk 3]
```diff
@@ -235,3 +239,216 @@ func testTrackerCreate(t *testing.T, tracker multiparts.Tracker) {
...
+func runKVPostgresBenchmark(b *testing.B, ctx context.Context, bmFunc RunBenchmarkFunc, concurrencyParams ...int) {
+	b.Helper()
+	store := kvtest.MakeStoreByName("postgres", databaseURI)(b, context.Background())
+	defer store.Close()
+	tracker := multiparts.NewTracker(kv.StoreMessage{Store: store})
+	bmFunc(b, ctx, tracker, concurrencyParams...)
+}
+
+func benchmarkCreateOp(b *testing.B, ctx context.Context, tracker multiparts.Tracker, concurrencyParams ...int) {
+	keys := generateRandomKeys(b.N)
```
Comment (nopcoder) [inline pkg/gateway/multiparts/tracker_test.go:316]: keys create base on N will be unexpected

**Review by nopcoder — APPROVED** [review]
lgtm except the part that create is based on the bench iteration which it will not serve the purpose

File: pkg/gateway/multiparts/tracker_test.go [hunk 4]
```diff
@@ -235,3 +239,216 @@ func testTrackerCreate(t *testing.T, tracker multiparts.Tracker) {
...
+				runSingleTrackerFlow(b, ctx, tracker, concurrencyParams[ConcurrencyNumParts])
+			}
+		}()
+	}
+
+	b.ResetTimer()
+
+	for n := 0; n < b.N; n++ {
+		ch <- true
+	}
```
[the remark this answers is not in the document]
Comment (itai-david) [reply pkg/gateway/multiparts/tracker_test.go:407]: Hi @arielshaqed. Thank you for that clarification.
Indeed, the current benchmark checks multiple routines, each running on a different file and thus, accessing a different entry in the KV Table.
In addition, the problem you are pointing in your comment, is indeed real.
I opened #3397 to follow up on that - hope that is OK with you

File: pkg/gateway/multiparts/tracker_test.go (hunk 3 above)
Comment (itai-david) [reply pkg/gateway/multiparts/tracker_test.go:316]: Opened #3387 to fix this
